firebase-installations/src/main/java/com/google/firebase/installations/FirebaseInstallations.java

```diff
@@ -103,4 +124,105 @@ String getApplicationId() {
+   *         check if cache empty or cache status is REGISTER_ERROR
+   *                                 |
+   *                           create random fid
+   *                                 |
+   *               update cache with cache status UNREGISTERED
+   *                                 |
+   *                           send http request to backend
+   *                                 |                  |
+   *  on success: set cache entry status to REGISTERED  |
+   *                                 |                  |
+   *                                 |         on failure: set cache entry status to REGISTER_ERROR
+   *                                 |                     |
+   *                                return          throw exception
+   *
+   *
+   *                      else if cached FID exists
+   *                             |            |
+   *     if cache status is UNREGISTERED      |
+   *                  |                   return cached FID
+   *     send http request to backend
+   *               |               |
+   *  on success: set cache        \
+   *  entry status to REGISTERED   \
+   *                |              \
+   *                |         on failure: set cache entry status to REGISTER_ERROR
+   *                |                     |
+   *              return          throw exception
+   * </pre>
+   */
+  @WorkerThread
+  private String createFirebaseInstallationId() throws FirebaseInstallationsException {
+
+    FiidCacheEntryValue cacheEntryValue = localCache.readCacheEntryValue();
+
+    if (cacheEntryValue == null
+        || cacheEntryValue.getCacheStatus() == FiidCache.CacheStatus.REGISTER_ERROR) {
+      String fid = Utils.createRandomFid();
+
+      boolean firstUpdateCacheResult =
+          localCache.insertOrUpdateCacheEntry(
+              FiidCacheEntryValue.create(fid, FiidCache.CacheStatus.UNREGISTERED, "", "", 0, 0));
+
+      if (!firstUpdateCacheResult) {
+        throw new FirebaseInstallationsException(
+            "Failed to update client side cache.",
+            FirebaseInstallationsException.Status.CLIENT_ERROR);
+      }
+
+      registerAndSaveFID(fid);
+
+      return fid;
+    } else if (!cacheEntryValue.getFirebaseInstallationId().isEmpty()) {
+
+      if (cacheEntryValue.getCacheStatus() == FiidCache.CacheStatus.UNREGISTERED) {
+        registerAndSaveFID(cacheEntryValue.getFirebaseInstallationId());
+      }
+
+      return cacheEntryValue.getFirebaseInstallationId();
+    }
+    return null;
+  }
+
+  /**
+   * Registers the created FID with FIS Servers if the Network is available and update the cache.
+   */
+  private void registerAndSaveFID(String fid) throws FirebaseInstallationsException {
+    try {
+      if (Utils.isNetworkAvailable(firebaseApp.getApplicationContext())) {
+        long creationTime = Utils.getCurrentTimeInSeconds();
+
+        InstallationResponse installationResponse =
+            serviceClient.createFirebaseInstallation(
```

Comment:
I think, I mentioned this before (I know it's pretty nitty):
Can we change the order of the parameters maybe?
like first FID, then API-Key, then all the metadata?
Or first the API-Key, then FID, then all the metadata?

Reply:
Changed it in the parent PR. https://github.com/firebase/firebase-android-sdk/pull/676/commits/34848c6d152bac0a13efdbaef4a341409735000d